Ammo Physics (3D Physics Extension)

Brackets-Coder opened the pull request:
### Ammo Physics
A physics extension based on the ammo.js physics library, which is a direct port of the world-renowned C++ Bullet Physics SDK.

The goal is to provide feature-complete, advanced, and performant 3D physics in a simple-to-understand manner. This extension aims to be consistent with Box2D and Simple3D.

https://github.com/user-attachments/assets/ffaa1071-dea3-4bf6-b804-900e788113a1

Task list:

- [x] Shapes
    - [x] Boxes
    - [x] Spheres
    - [x] Cylinders
    - [x] Cones
    - [x] Capsules
    - [x] Convex Hulls
    - [x] Triangle Meshes
    - [x] Compound bodies
    - There are some others (planes, heightfields), but I don't want to support these
- [x] Physical Materials (friction, restitution, etc.)
- ~~[ ] Constraints (All types)~~
- [x] Collision:
    - [x] Collision Detection
    - [x] Raycasting
    - [x] Enable/disable collision response
- [x] Manual Impact forces (like "push with force" block in Box2D)
- [x] Bug fixing

Things I'm considering now or for a later update:
- Vehicle support
- ~~Debug rendering support~~
- ~~Soft Bodies~~

Miscellaneous Notes Mainly for Moderators:
- **_ChatGPT_** was used to **_assist_** and **_enhance_** the making of this extension, simply because I couldn't find good quality, coherent documentation of Ammo.js, and attempting to figure out how to use it via the developer inspector isn't fun. **_It did not write all of the code for me and I am still the main author of this extension._** It just helped me learn Ammo.js. Hope this doesn't cause conflicts.
- [x] ~~I'm not good at graphic design so if someone wants to make a banner/thumbnail that'd be great~~
- [x] ~~Still have to incorporate `Scratch.Cast` where necessary, right now I'm just working on functionality though~~

Put an emoji (👍, 🚀) if you like this extension

Brackets-Coder committed:
Create Ammo Physics

Brackets-Coder committed:
Create AmmoPhysics.js

Brackets-Coder commented on extensions/extensions.json:
```diff
@@ -4,6 +4,7 @@
   "stretch",
   "gamepad",
   "box2d",
+  "AmmoPhysics",
```
@CubesterYT 
Should I leave this where it is or move it closer to Simple3D? I'm fine with either, but I don't think it fits around all of the other rendering extensions because it doesn't render anything. At the same time, it's 3D, so it can go by Simple3D, but it's also Physics, so it kinda makes more sense to be by Box2D, but it doesn't necessarily have to be at the top of the gallery.

CubesterYT replied:
Put it near Box2D maybe

WAYLIVES commented:
Wow! It's very cool and useful for people like me)

Brackets-Coder commented:
Added partial support compound bodies - all child shape types.
Also working on convex and concave mesh support... it'll be a while.

WAYLIVES commented:
wow

Brackets-Coder committed:
Merge branch 'TurboWarp:master' into ammo-physics

Brackets-Coder committed:
Merge branch 'TurboWarp:master' into ammo-physics

Brackets-Coder commented:
Finally got convex hulls working 🥳 🎉 

https://github.com/user-attachments/assets/ad8d1814-cfd9-4ece-a145-18e206262783

Still have to optimize and bug fix everything and add more features

Brackets-Coder commented:
@Xeltalliv 

Sorry to ping you, but I've got a quick question. I'm trying to add concave triangle mesh support but I can't decide how to go about parsing vertex and face data. I want to have a block that can do vertices and/or faces directly from a list and I want another block that can do OBJ files, but then it gets tricky because I have to have use cases for handling quads and n-gons when Ammo.js requires that all faces be triangles. Does Simple3D internally manage quads and n-gons or do you triangulate the mesh before uploading? If so, how difficult would it be to implement mesh triangulation from scratch, and which methods should I use? I was considering adding various triangulation methods or having a library do it for me, but wouldn't it be easier (albeit less user-friendly) to have it only accept OBJ files that are already triangulated (can't have more than 3 vertices per face)? E.g., require that users triangulate their meshes in Blender or something like that before importing

Brackets-Coder commented:
> It's a bit complicated. Simple3D itself can have meshes represented in 3 ways: triangles, triangle strips and triangle fans. Positions of each of the vertices can either be listed directly, or be specified once per unique position and referenced multiple times from the list of indices, to share one vertex between multiple primitives. Triangles require 3 vertices per triangle. Triangle strips and fans require 3 vertices for the first triangle and only 1 vertex for each subsequent triangle, where 2 old vertices get reused. In case of triangle strip it reused 2 immediately previous vertices. In case of triangle fan it reuses 1 previous vertex and the 1st vertex. Index of 0 can be used as a separator to have multiple unconnected triangle strips/fans. ![image](https://private-user-images.githubusercontent.com/90340988/431477806-fb021e0a-c235-4d42-8ee5-c400cea2c676.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxMzMyMDYsIm5iZiI6MTc0NDEzMjkwNiwicGF0aCI6Ii85MDM0MDk4OC80MzE0Nzc4MDYtZmIwMjFlMGEtYzIzNS00ZDQyLThlZTUtYzQwMGNlYTJjNjc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA4VDE3MjE0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRkYzQwZWU5OTU1ZDVhYzFhNTBkMTE0OTI1N2MwNjAwYTA3OTQwNzQwYTk1ZWQ2MzYxYTMzN2Q3Mzk5ODlkNTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.kzu31nvklVkBf5hYLHVK6blmH-th6vRcocNwXWtK55c)
> 
> The built-in OBJ importer doesn't use any of those modes and always just outputs a simple list of triangles pre-processed using fan triangulation algorithm. It does not handle concave n-gons.

I'm aware of the multiple different primitive types, that's partially why I was asking. Thanks for the help

Brackets-Coder committed:
Raycasting Rewrite and Scratch.Cast

Brackets-Coder commented:
### Progress Update

- Rewrote the entire raycasting system to be more efficient/capable and lining up with Scratch best practices (still have to add icons)
- Added Scratch.Cast to every block argument of every block.

@CubesterYT I know this isn't near ready for review yet... but whenever you can could you please take a quick look at the casting to make sure I did it right?

Brackets-Coder commented:
@CubesterYT 

You might want to hold off, I think I broke something in the last commit; convex hulls are clipping through bodies.

Brackets-Coder committed:
Remove cast because yeah

Brackets-Coder committed:
Fix issues

Brackets-Coder commented:
I just removed the casting because I think that was what was causing the problems. Also don't know why my last commit passed ESLint but now it's warning me about things that should've thrown an error. I fixed those so now we should be good moving forward.

I have to add mesh support for `btBvhTriangleMeshShape` and `btGImpactMeshShape` and constraints and impulse forces. Probably should also add vehicles and some kind of player management.

Brackets-Coder commented:
### Update
Just got some form of meshes partially supported:

https://github.com/user-attachments/assets/94322ceb-a24e-490d-bbb2-563e84e9ce3b

Still have to clean it up, optimize and things. Only the dynamic (but slower) `btGImpactMeshShape` works right now, for some reason the static `btBvhTriangleMeshShape` doesn't work and clips through other bodies.

Right now it only supports meshes where the data is from a Scratch list like this:
```
-1 -1 -1
0 0 0
1 1 1
```

Where each item contains three separated coordinates and every three items corresponds to the vertices of one triangle. this means that all vertices must be in order. There are multiple delimiter options available in the RegEx that splices this. I have not yet added support for vertex indexing, but that's planned before release. Vertex indexing will also make it easier to implement OBJ file processing, but OBJ files must be triangulated before importing into the Scratch lists as Ammo.js expects triangulated, planar meshes with no n-gons or non-manifold geometry.

I also removed the casting and fixed the convex hulls.

#### Task list:

- [ ] Finish Meshes
- [ ] Add all constraint types
- [ ] Add impulse forces
- [ ] Bug fix
- [ ] Best practices
- [ ] Optimize
- [ ] Maybe add vehicles and player management?

Brackets-Coder commented:
<img width="1728" alt="Screenshot 2025-04-10 at 1 41 14 PM" src="https://github.com/user-attachments/assets/1fe87111-9400-432e-b57f-e81d4e64fbcc" />
I'm going to have to fix that.

Brackets-Coder commented:
They're very glitchy and there's a lot of tunneling when two impact meshes intersect...

Brackets-Coder commented:
https://github.com/user-attachments/assets/06356cae-f610-4c11-a2f5-070001e89d29

@Xeltalliv Since you have experience writing physics engines, do you have any idea how to fix this? I've tried various things like changing maxSubSteps, simulation framerate, project framerate, etc. but nothing works. It only occurs when one or more `btGImpactMeshes` intersect with each other. One mesh and one convex hull works as expected. Multiple convex hulls work as expected. I've looked at demos that showcase multiple meshes in a chain with a wrecking ball at the bottom and it works fine so I know it's not a library limitation.

I could try using a more specialized collision algorithm but I don't want to sacrifice performance for simpler simulations.

Brackets-Coder commented:
Also, Ammo.js seems to support wireframe debug rendering. I might want to add a layer to the canvas that can show contacts, convex hulls, velocities, etc.

Xeltalliv commented:
This looks like a faulty collision detection. Changing framerate, sub-steps and solver iterations is mostly useful when you have a lot of bodies interacting in a chain reaction, which isn't the case here.

> I could try using a more specialized collision algorithm but I don't want to sacrifice performance for simpler simulations.

Not sure exactly what you are referring to, because there isn't much documentation, but I asked ChatGPT how to correctly use `btGImpactMeshShape` and it on it's own mentioned that using `btGImpactCollisionAlgorithm::registerAlgorithm` is absolutely mandatory for it to work correctly.

Brackets-Coder commented:
> This looks like a faulty collision detection. Changing framerate, sub-steps and solver iterations is mostly useful when you have a lot of bodies interacting in a chain reaction, which isn't the case here.
> 
> > I could try using a more specialized collision algorithm but I don't want to sacrifice performance for simpler simulations.
> 
> Not sure exactly what you are referring to, because there isn't much documentation, but I asked ChatGPT how to correctly use `btGImpactMeshShape` and it on it's own mentioned that using `btGImpactCollisionAlgorithm::registerAlgorithm` is absolutely mandatory for it to work correctly.

I'm already registering the GImpactCollisionAlgorithm, but ChatGPT suggests using a more narrow phase parent collision algorithm I don't really think that will affect anything. Looking at the official demos, `btDbvtBroadphase` is being used and that's what I'm using. I could try to see if I'm registering the impact collision algorithm correctly, but I think I am:

```
    let dispatcher = new Ammo.btCollisionDispatcher(collisionConfig);
    Ammo.btGImpactCollisionAlgorithm.prototype.registerAlgorithm(dispatcher);
```

Brackets-Coder commented:
Now it's telling me that I can't do anything about it and to accept the limitations but I know for sure it's possible. There's like twelve different GImpactMeshes all interacting here:

https://github.com/user-attachments/assets/00c55cdf-77fb-4b02-90d8-66466692e838

Brackets-Coder commented:
I think my triangle solver is faulty and ChatGPT is too dumb to help me.

Brackets-Coder commented:
Working on manual forces (impulses, pushes, etc. like in Box2D), cleaning raycasting rewrite, fixing some collision filtering bugs when attempting to disable collision response, etc...

I'm taking a break on triangle meshes rn until I can figure out what in the world is going wrong

Brackets-Coder commented:
So I've been busy with life for a bit and AFK so this PR is gonna be delayed

Brackets-Coder commented:
@CubesterYT how unadvisable is it to provide console warnings for user debugging in extensions?

Brackets-Coder commented:
Added push forces, still have to add torque and clear forces but those won't take long at all

Brackets-Coder commented:
Y'all I finally got triangle meshes working... it was a face indices issue. Just have to bug fix it a bit when using invalid lists... but right now alpha support works great! You can see 3 meshes interacting with concave collisions in this video:

https://github.com/user-attachments/assets/842713de-a2c9-4fbb-baca-34b38cfe27e5

Brackets-Coder commented:
Added basic constraint support, I've tested ball constraint and fixed constraint and both seem to be somewhat working.
However, I seem to have a larger issue regarding the memory management of objects when pressing the green flag, often it crashes or freezes the physics world.

Sorry for the lack of a video, but hey - they work I guess

Brackets-Coder committed:
Add triangle meshes and constraints

- Triangle meshes
- Constraints
- Manually Applied Forces
- Set body gravity
- Reset world block
- Raycast Icon
- Improved naming structure of some blocks

Brackets-Coder commented:
Just fixed collision response block not working - the task list is now complete, and although the first version of the extension is "feature-complete," there's still a few more features I'd like to add in future version after the initial merge and more bugs to fix before that can happen.

Brackets-Coder commented:
@CubesterYT 

I believe that @GarboMuffin once said to @yuri-kiss that "mutating args is a bad smell". Is this okay at the very top of the block for `Scratch.Cast`ing and code readability purposes?

Brackets-Coder committed:
v0.9-beta

- New Features
    - Added safe object replacement for objects
    - Added mesh triangulation validation, now won't accept non-triangulated meshes
    - Added support for `btBvhTriangleMeshShape` - you can now use BVH accelerated static meshes
    - Added a block that returns the name of every existing body in a comma-delimited list
- Bug Fixes
    - Improved memory management to make the editor-crashing `Aborted: Out of Memory` issue happen less often
    - Fixed the toggle collision response block not working, now works as expected - you can make "ghost" bodies if you like
    - Fixed the set body gravity block not working
    - Fixed the change body rotation block not working as expected, now consistent with set body rotation
    - Fixed an issue where some triangle data was parsed incorrectly by subtracting one from the index
    - Fixed the push forces blocks from not working when applied to a body that fell asleep

Brackets-Coder committed:
Create AmmoPhysics.md

Brackets-Coder committed:
Merge branch 'TurboWarp:master' into ammo-physics

Brackets-Coder committed:
Fixes and updates

Brackets-Coder committed:
Raycasting and Forces documentation

Kreerio commented:
kk

Brackets-Coder committed:
Some documentation improvements

Brackets-Coder commented:
there's no freaking documentation writing constraints is a pain 😞

Brackets-Coder commented:
@CubesterYT I'm pretty sure this isn't possible but the best ammo.js projects use WebAssembly / Web workers to let the physics run async in the background so it doesn't hinder rendering processes, etc, also have more memory capacity. 

I'm assuming this isn't possible in TW and it has to be native JS?

Brackets-Coder committed:
Create AmmoPhysics.svg placeholder banner

Brackets-Coder committed:
Fix step not working at 0 FPS and comments

- Improves some of the comments or expands on them for self future reference
- Fixes an issue where the step simulation block would not behave correctly at 0 FPS, it now implies the screen refresh rate as Turbowarp does

Brackets-Coder committed:
Still working on documentation...

Brackets-Coder commented:
@GarboMuffin and/or @CubesterYT 

I'd like this extension to be able to support WebAssembly as it has much higher performance and much lower memory limitations.
The way I've set it up now is that a ammo.wasm.wasm file would be hosted at extensions.turbowarp.org/MasterMath/ammo.wasm.wasm

The extension would check if that file exists, if so load the WASM ammo with it, if not fall back to the default offline native JS version (at the cost of performance). This way there's no external domains and the extension should always load with some kind of Ammo.

In practice, it looks like this:
<img width="1840" height="1195" alt="Screenshot 2025-09-18 at 8 44 10 PM" src="https://github.com/user-attachments/assets/76f3fcff-2446-47d7-8f5a-870eed23762e" />

would something like this even work?

GarboMuffin commented:
I'll just be real with you. The likelihood that I would tolerate merging a 2MB extension is really low right now

Brackets-Coder commented:
> I'll just be real with you. The likelihood that I would tolerate merging a 2MB extension is really low right now

Alright, I'll just use the native version and use Terser to minify it compressed. I already knew it was really big but I don't know that I can do much to get it smaller...

Brackets-Coder commented:
Alternatively, I could build my own version of Ammo without the extra stuff that is much smaller than the prebuilt library, but it would probably remove things like the wireframe debug drawer which might be nice (but a lot of work and size) to implement later...

Brackets-Coder commented:
@GarboMuffin 
I made a custom build of Ammo.js with soft bodies and debug drawing disabled. I've also enabled WebAssembly and embedded it directly into the JS file, so I don't need to access extra files. 

This makes it run faster, have "infinite" memory limitations, and the file size is now only 1.2 MB. Happy?

Brackets-Coder committed:
Use custom build

- WebAssembly Embedded into JS for higher performance and scalable memory (Fixes the Aborted: Out of Memory error)
- Remove soft bodies and debug drawer for half the size
- And something else IDK

Brackets-Coder committed:
ummm format passed why is it failing

Brackets-Coder commented:
@CubesterYT since the last commit where I made a custom build of the library, format is failing even though `npm run check-format` passed on my local machine. Why is that?

GarboMuffin commented:
it's because of the prettier update pull request you merged where now your local prettier doesn't match what's on your pull request branch

that's why the pull request wasn't merged for a long time

rebase, reinstall, reformat, itll go away

Brackets-Coder commented:
> it's because of the prettier update pull request you merged where now your local prettier doesn't match what's on your pull request branch
> 
> that's why the pull request wasn't merged for a long time
> 
> rebase, reinstall, reformat, itll go away

Oops sorry just trying to help

Brackets-Coder committed:
Merge branch 'TurboWarp:master' into ammo-physics

Daniel-Dami commented:
this extension is super cool!!

Brackets-Coder committed:
Reduce size and fix prettier

Brackets-Coder commented:
Previous commit reduces the file size ≈0.2 MB
The extension is now only 1MB, I don't think I can do much better than that.

Brackets-Coder commented:
> Previous commit reduces the file size ≈0.2 MB
> The extension is now only 1MB (1020 KB, around half the original size); I don't think I can do much better than that without > breaking functionality.
>> that's really great!

@CubesterYT ["The weight loss update"](https://github.com/TurboWarp/extensions/pull/2254#issuecomment-3343842355) lol

Brackets-Coder commented:
> @GarboMuffin
> 
> Hello, I made a custom build of Ammo.js with soft bodies and debug drawing disabled. I've also enabled WebAssembly and embedded it _directly_ into the JS file, so I don't need to access extra files on the server, which probably wouldn't work anyway. I also configured the build to default at 128 MB memory limit and dynamically increase if needed. Terser minified and compressed it.
> 
> This makes it run faster, have "infinite" memory limitations (can be both good or bad, but it fixes the crashing Out of Memory issue), and the entire extension size is now only 1.2 MB (with all the blocks and stuff too). Is that good with you or should I compress further? What is the preferred size? I could probably get it a few hundred kilobytes, but that would be more aggressive and I don't think I can go further than that.

Brackets-Coder commented:
Honestly I think I should just crank out getting constraints figured out and finishing documentation

If all goes well this can be merged but there are several concerns that need to be addressed

Brackets-Coder committed:
Prepare for release

Brackets-Coder committed:
update format

Brackets-Coder committed:
Fix a few invalidation issues with the meshes

Brackets-Coder committed:
Remove extra console logs

Brackets-Coder committed:
Fix meshes and add body active

Brackets-Coder committed:
Update AmmoPhysics.svg

Brackets-Coder committed:
Fix image

Brackets-Coder committed:
Merge branch 'TurboWarp:master' into ammo-physics

Brackets-Coder commented:
<img width="960" height="267" alt="Screenshot 2025-10-09 at 8" src="https://github.com/user-attachments/assets/28c76107-5099-4906-ba3f-af178b6f9a22" />

@GarboMuffin and/or @CubesterYT after merging upstream I'm getting this error when doing `npm run dev` EDIT: the same issue is happening for `validate` too
Anyone have any idea what's going on

Brackets-Coder committed:
Fix slash notation, cast body active

Brackets-Coder committed:
Fix casting and body active

Brackets-Coder commented:
Oops meshes being weird

Brackets-Coder commented:
https://github.com/user-attachments/assets/091eafa2-00cd-4292-bd0a-910bea17e006

Yeah I think it's working

Brackets-Coder committed:
Merge branch 'TurboWarp:master' into ammo-physics

MasterThomas15 commented:
Hey, why isnt there any bounce?

MasterThomas15 commented:
nvm restuision

MasterThomas15 commented:
... but in the extension mass doesnt do anything

MasterThomas15 commented:
On new ver, forces dont work at all (they give error) TypeError: bodies[name][force] is not a function

wiktorlaskowski commented:
> @wiktorlaskowski and @WAYLIVES might want to take a look at the placeholder banner and see if it's acceptable

I?!

Brackets-Coder commented:
> On new ver, forces dont work at all (they give error) TypeError: bodies[name][force] is not a function

Thank you for finding this, it's reproducible on my machine. It was a simple variable naming/reassignment overlook that took me 30 seconds to fix, so it will be resolved in the next commit.

Brackets-Coder committed:
Fix the forces not working and add a few things

Brackets-Coder committed:
Merge branch 'TurboWarp:master' into ammo-physics

Brackets-Coder commented:
Hello @GarboMuffin,
Now that external dependencies are possible, I'd like to include it in this extension to cut its size. However, the utilization of a custom build (for WASM embedding), could I host the custom build here in this repository and fetch it from the raw Github URL, or must I rely on other alternatives?

Brackets-Coder committed:
Remove library and use dependency import

Brackets-Coder commented:
The previous commit uses the new ESM stuff to do `Scratch.external` importing instead of pasting the entire libraries, this reduces the file size by like 89% (1000KB —> 108 KB) and makes it much easier to read

Tradeoff is that it doesn't seem to work in my version of the Desktop editor, which I don't think has been updated yet?

wiktorlaskowski commented:
> > > @wiktorlaskowski and @WAYLIVES might want to take a look at the placeholder banner and see if it's acceptable
> > 
> > 
> > I?!
> 
> Not sure what you mean by this

I mean I will need to figure this out?
Hint, no I don't even know JS I just use AI for everything for now. Well apart from thumbnails for those extensions, I made them myself.

Brackets-Coder commented:
> > > > @wiktorlaskowski and @WAYLIVES might want to take a look at the placeholder banner and see if it's acceptable
> > > 
> > > 
> > > I?!
> > 
> > 
> > Not sure what you mean by this
> 
> I mean I will need to figure this out? Hint, no I don't even know JS I just use AI for everything for now. Well apart from thumbnails for those extensions, I made them myself.

I just want to know if you think the banner I made is okay, I don't need you to review the extension it's alright

Brackets-Coder commented:
I'm not normally one to do this but I think it'd be nice if I could have everyone review this PR just because it's so big
I don't care how much time you take, I don't care if you don't do it, I understand it likely won't be merged, but as many eyes as possible would be very helpful.
Thanks so much everyone!

Brackets-Coder committed:
this should fix validate

Brackets-Coder commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,2584 @@
+        } else {
+          return;
+        }
+      }
```
Is this mess of code acceptable? Are there potential loopholes where my validation checks would fail and create an invalidated mesh body? I've done a lot of testing and haven't found anything...

Brackets-Coder commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,2584 @@
+      });
+      //* ------------
+
+      runtime.on("PROJECT_START", () => {
```
This is here as a memory management tool. Will some users find it unintuitive if the extension automatically clears all bodies, rays, etc. and resets the world when the green flag is pressed, especially considering that there's a "reset world" block?

Brackets-Coder commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,2584 @@
+        rays = {};
+      });
+
+      // These SVG Icons from Blender source code: https://github.com/blender/blender/tree/main/release/datafiles/icons_svg
```
Inquiring whether the use of third party icons requires license attribution; I have included a link here.

Brackets-Coder commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,2584 @@
+        }
+
+        setGravity({ x, y, z }) {
+          world.setGravity(new Ammo.btVector3(x, y, z));
```
I may have forgotten to cast these, I don't know why I didn't?

GarboMuffin commented:
that would mean your vm is outdated

Meehdrescher commented:
> that would mean your vm is outdated

Is that actually a possible issue, even if I just cloned everything into a completely new environment about 10 minutes ago?

Brackets-Coder committed:
Merge branch 'TurboWarp:master' into ammo-physics

Brackets-Coder commented:
@Civ3ro pointed out an issue with the step simulation rate. It runs far too frequently in a forever loop in which there is nothing being rendered. Putting in other stage-sensitive blocks (like draw calls or even a "move steps" block) mitigates the issue entirely. 

I'll continue looking into this

SharkPool-SP commented:
I could review this in a couple weeks or so, whenever I get my big haul of work out of the way

Brackets-Coder committed:
Misc Updates and Button Folders

Brackets-Coder commented:
@SharkPool-SP Do you mind returning the favor? 
It's been "a couple weeks" 😆
I totally get it if you need more time though

Brackets-Coder commented:
> I'm probably going to abandon my duties here at some point.

<img width="430" height="189" alt="image" src="https://github.com/user-attachments/assets/256a699e-d350-49ac-a70f-42cd9d992a96" />

EDIT: gif don't work in the comment section oof

SharkPool-SP commented:
> @SharkPool-SP Do you mind returning the favor? It's been "a couple weeks" 😆 I totally get it if you need more time though

after homework we'll see about today

Brackets-Coder commented:
> > @SharkPool-SP Do you mind returning the favor? It's been "a couple weeks" 😆 I totally get it if you need more time though
> 
> after homework we'll see about today

Alright thx

Brackets-Coder committed:
Documentation revision & corrections

SharkPool-SP commented:
> > > @SharkPool-SP Do you mind returning the favor? It's been "a couple weeks" 😆 I totally get it if you need more time though
> > 
> > 
> > after homework we'll see about today
> 
> Alright thx

got carried away, will do today 👍

Brackets-Coder commented:
> got carried away, will do today 👍

Okay thank you. Just fixed a bunch of grammar issues in the docs so hopefully that'll be a bit easier on you

Brackets-Coder committed:
fix typescript warnings

Brackets-Coder committed:
oops left that on the bottom

SharkPool-SP commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,2688 @@
+        const quaternion = new Quaternion(q.w(), q.x(), q.y(), q.z());
+        const euler = quaternion.toEuler("XYZ");
+        return {
+          x: euler[0] * (180 / Math.PI),
```
you calculate this multiple times, you can just precalculate it an use the value whenever. Its slightly more optimized then doing this math operation >3 times

SharkPool-SP commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,2688 @@
+      function eulerToQuaternion(x, y, z) {
+        // @ts-ignore
+        let quaternion = Quaternion.fromEuler(
+          x * (Math.PI / 180),
```
you calculate this multiple times, you can just precalculate it an use the value whenever. Its slightly more optimized then doing this math operation >3 times

SharkPool-SP commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,2688 @@
+      let autoReset = true;
+
+      runtime.on("PROJECT_START", () => {
+        if (autoReset) resetWorld();
```
ideally should reset world on project stop too no?

SharkPool-SP commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,2688 @@
+            forces: false,
+          };
+
+          this.reset = () => {
```
this is a bit of a broad function name, perhaps refreshPallete or something

SharkPool-SP commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,2688 @@
+      function resetWorld() {
+        world.setGravity(new Ammo.btVector3(0, -9.81, 0));
+        for (const key in bodies) {
+          if (Object.prototype.hasOwnProperty.call(bodies, key)) {
```
if you care about optimization, dont use regular objects (mutable by the user) like `{}` with hasOwnProperty.

You could just use `Object.create(null)` for these types of objects, then you dont need to check hasOwnProperty since theres no reserved keys. Maps are also an option, they are pretty optimized

SharkPool-SP reviewed (changes requested):
review numero uno

SharkPool-SP commented:
I havent tested functionality yet, this is just code conventions

Brackets-Coder commented:
Thanks. Let me know if you'd like a sample project.
I'll be addressing your concerns shortly

Brackets-Coder committed:
add suggested edits

Brackets-Coder commented:
[Ammo Stress Test.zip](https://github.com/user-attachments/files/24600760/Ammo.Stress.Test.zip)

That should suffice for testing. It doesn't have every feature but renaming it to .sb3 should give you a pretty good foundation. 

Let me know how it goes, thanks so much!

Brackets-Coder committed:
Documentation revision & Missing Blocks

Brackets-Coder committed:
Remove activate body and reword block text

Brackets-Coder commented:
!format

Brackets-Coder committed:
Bro why

SharkPool-SP commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,2697 @@
+        }
+
+        allBodies() {
+          return Cast.toString(Object.keys(bodies));
```
shouldnt this stringify the array

Brackets-Coder replied:
currently reports something like `0,1,2,3,4,5,6,7,8,9,ground`

I can make it use JSON.stringify instead if that's what you're asking

SharkPool-SP replied:
Yeah do that since all other reporters like that typically return it as json

SharkPool-SP reviewed (changes requested):
error
<img width="1114" height="220" alt="Screenshot 2026-01-14 at 1 07 02 AM" src="https://github.com/user-attachments/assets/849cca6e-2314-4878-943e-74927c9b432f" />

SharkPool-SP commented:
you might also want to go the boxed physics route and allow boxing bodies within the stage bounds.

Also what was it you wanted with delta time?

Brackets-Coder commented:
> error

I can't reproduce this. Do you know what caused the error or what you did to make it show up? There shouldn't even be a `not a function` error, the menu is static so always maps to an existing function

> You might also want to go the boxed physics route and allow boxing bodies within the stage bounds

It's not practical to map screen space to world space boundaries in Ammo.js, and seeing as most people would use this in their 3D worlds I can't see that being used for more than static top-down or side games. If they want that, they can calculate the transformation themselves and make thin boxes covering the bounds of the camera frustum

> Also what was it you wanted with delta time?

There's a small glitch where if there's no draw calls to the stage (no sprite updates or pen blocks) ammo physics speeds up dramatically. I'm assuming it's because it's refusing to sync to the frame loop. Either way, it's a negligible issue because this extension is always meant to be used with visual feedback

Brackets-Coder committed:
Add debugging + minor patch

Adds debugging to the documentation and a parameter to refreshBlocks to prevent the refreshing of the entire palette

Brackets-Coder commented:
!format

Brackets-Coder committed:
bruh

Brackets-Coder committed:
requested change

Brackets-Coder committed:
oops

Brackets-Coder commented:
@SharkPool-SP I know you're probably no where close to approving or merging but I'd like to add a few body utility blocks before that happens, so just don't merge yet

I'll try and be quick

Brackets-Coder commented:
Alright that was quick. I added them, now I just have to test them

Brackets-Coder committed:
Add extra body utility blocks

SharkPool-SP commented:
> @SharkPool-SP I know you're probably no where close to approving or merging but I'd like to add a few body utility blocks before that happens, so just don't merge yet
> 
> I'll try and be quick

its not like I can merge, I need another approval

SharkPool-SP commented:
> > error
> 
> I can't reproduce this. Do you know what caused the error or what you did to make it show up? There shouldn't even be a `not a function` error, the menu is static so always maps to an existing function
> 
> > You might also want to go the boxed physics route and allow boxing bodies within the stage bounds
> 
> It's not practical to map screen space to world space boundaries in Ammo.js, and seeing as most people would use this in their 3D worlds I can't see that being used for more than static top-down or side games. If they want that, they can calculate the transformation themselves and make thin boxes covering the bounds of the camera frustum
> 
> > Also what was it you wanted with delta time?
> 
> There's a small glitch where if there's no draw calls to the stage (no sprite updates or pen blocks) ammo physics speeds up dramatically. I'm assuming it's because it's refusing to sync to the frame loop. Either way, it's a negligible issue because this extension is always meant to be used with visual feedback

code was literally just
<img width="679" height="209" alt="image" src="https://github.com/user-attachments/assets/ff7f0d6a-3c46-49d2-8e09-5c70ac046671" />

Brackets-Coder commented:
> its not like I can merge, I need another approval

Yeah of course, I thought adding the blocks might take days, not 45 minutes

Brackets-Coder commented:
> code was literally just

Found the problem:
```javascript
                force: {
                  type: Scratch.ArgumentType.STRING,
                  defaultValue: "force",
                  menu: "forceMenu",
                },
```

Default value did not match a function name, will remove and fix now

Brackets-Coder committed:
fix force issue

Brackets-Coder commented:
@SharkPool-SP might wanna test again? Everything should be working now
Let me know if I can do anything else for you

SharkPool-SP commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,2882 @@
+  // @ts-ignore
+  Ammo()
+    .then(function (Ammo) {
+      console.log(Ammo);
```
this can be removed

Brackets-Coder replied:
Okay, but it was making a warning that wasn't a problem and I simply wanted it to go away. I'll remove it

SharkPool-SP replied:
What?

SharkPool-SP commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,2882 @@
+          return JSON.stringify(Object.keys(bodies));
+        }
+
+        createBoxBody({ name, mass, x, y, z }) {
```
just realized this is gonna be problematic for all your blocks.

Its not good practice to do this `func({ arguments }) {`, typically you should just use  `func(args) {`,.

I say this because there are multiple arguments in your block, and for many of them, this included, you dont cast your values.

x, y, z, are not converted to numbers, as with multiple other blocks.

Brackets-Coder replied:
Other than the lack of casting, what's wrong with func({ arguments})? other extensions do it too

Doing it that way makes it easier to see what I'm working with in my opinion

SharkPool-SP replied:
I don't know, it's just not something modern extensions do from what I know 🤷‍♂️

Brackets-Coder replied:
Is it enough of a problem to justify refactoring every block function, or can I get away with just adding casting?

SharkPool-SP replied:
casting sure yeah

SharkPool-SP reviewed (changes requested):
besides that everything should be fine

Brackets-Coder commented:
> besides that everything should be fine

Okay

Brackets-Coder committed:
Everything should be formatted now

Brackets-Coder commented:
Everything should be casted now. I've done a quick run through of the blocks to make sure but please double-check as I might've missed something. 

Thanks so much for your help!

SharkPool-SP commented:
btw I didnt mean remove the //ts-ignore, i meant remove the console.log()

SharkPool-SP commented:
add that back and remove that console log and ill approve

Brackets-Coder committed:
Alrighty

Brackets-Coder commented:
Should be good to go, thanks for helping

Brackets-Coder commented:
Oops gotta add docs for the blocks I added yesterday

Brackets-Coder commented:
@SharkPool-SP should this be JSON stringified as well?

      allBodiesTouchingBody({ body }) {
        return bodies[Cast.toString(body)]?.collisions;
      }

SharkPool-SP commented:
> @SharkPool-SP should this be JSON stringified as well?
> 
> ```
>   allBodiesTouchingBody({ body }) {
>     return bodies[Cast.toString(body)]?.collisions;
>   }
> ```

ideally any block that returns a list of something should be stringified, its more usable with extensions like JSON

Brackets-Coder committed:
Stringify and documentation revisions

Brackets-Coder commented:
alright everything should be ready now

Brackets-Coder committed:
raycasting utilities

Brackets-Coder committed:
add get velocity

Brackets-Coder commented:
!format
Me finding out some of the raycast blocks don't work:

Brackets-Coder commented:
@Xeltalliv Raycast with direction (pitch/yaw) doesn't seem to work properly. I've been able to replicate the expected result using point-to-point raycast and calculating the transformation manually using Simple3D, but not here?

```javascript
        rayCastDirection({ name, x, y, z, rotX, rotY, rotZ, distance }) {
          name = Cast.toString(name);
          if (rays[name]) {
            Ammo.destroy(rays[name]);
            delete rays[name];
          }
          const pitch = (Cast.toNumber(rotX) * Math.PI) / 180;
          const yaw = (Cast.toNumber(rotY) * Math.PI) / 180;
          const dir = new Ammo.btVector3(
            Math.cos(yaw) * Math.cos(pitch),
            Math.sin(pitch),
            Math.sin(yaw) * Math.cos(pitch)
          );
          dir.op_mul(distance);

          const from = new Ammo.btVector3(
            Cast.toNumber(x),
            Cast.toNumber(y),
            Cast.toNumber(z)
          );
          const to = new Ammo.btVector3(
            from.x() + dir.x(),
            from.y() + dir.y(),
            from.z() + dir.z()
          );

          const rayCallback = new Ammo.AllHitsRayResultCallback(from, to);
          world.rayTest(from, to, rayCallback);
          rays[name] = rayCallback;
          rays[name].endpoint = to;
        }
```

Brackets-Coder committed:
format

Brackets-Coder committed:
fix raycast direction not working

Brackets-Coder committed:
Documentation revision

Brackets-Coder committed:
update body touching ray naming & function

Brackets-Coder commented:
Should be ready for another review now that I've resolved the issues and updated documentation.

Here's the sample project (featuring raycasting and 50 convex hulls):
[AmmoPhysicsDemo.zip](https://github.com/user-attachments/files/24697960/AmmoPhysicsDemo.zip)

Doing a stress test, my computer can easily handle 250 Suzanne convex hulls. With 500 dynamic concave meshes, that goes down to about 1FPS, but it doesn't crash

I'll be requesting a review but I will be patient and respect your time. Take as much as you need

Brackets-Coder committed:
Update AmmoPhysics.md

Brackets-Coder commented:
@CubesterYT I have the unreasonable goal of trying to get as many of my PRs (assigned) merged as soon as possible, and probably at least one before the end of the month.

I understand if you're busy, but now that it's been a week, could you please give this a quick skim review and then do the web hook update so we can get #1873 merged? Thanks!

Brackets-Coder commented:
@PPPDUD are you still busy and will your review activity significantly decline in the foreseeable future?

PPPDUD commented:
> @PPPDUD are you still busy and will your review activity significantly decline in the foreseeable future?

I'm "busy" in the sense that I have things to do that I consider more important than TurboWarp, though that doesn't mean that I'm constantly working on stuff. As for my review activity, most likely yes.

Brackets-Coder commented:
> > @PPPDUD are you still busy and will your review activity significantly decline in the foreseeable future?
> 
> I'm "busy" in the sense that I have things to do that I consider more important than TurboWarp, though that doesn't mean that I'm constantly working on stuff. As for my review activity, most likely yes.

I see... so we'll likely need more eventually

PPPDUD commented:
> > > @PPPDUD are you still busy and will your review activity significantly decline in the foreseeable future?
> > 
> > 
> > I'm "busy" in the sense that I have things to do that I consider more important than TurboWarp, though that doesn't mean that I'm constantly working on stuff. As for my review activity, most likely yes.
> 
> I see... so we'll likely need more eventually

Yes.

Brackets-Coder commented:
@PPPDUD or @DogeisCut If you aren't busy next week, can either of you do a quick review?

Brackets-Coder commented:
@DogeisCut Quick skim if you don't mind?

Civ3ro commented:
Hey, I could not find an issue place, so I'll place it here:
<img width="712" height="388" alt="image" src="https://github.com/user-attachments/assets/4925bc00-cf41-4a8f-afc6-9a65de916bb5" />
I guess here:
```js
function processVertices(list) {
  const points = [];
  const array = list;
  if (array) {
    for (let i = 0; i < array.length; i++) {
      if (array[i] != "") {
        const item = array[i].split(" ");
        if (item.length !== 3) {
          return;
        }
        points.push(
          new Ammo.btVector3(
            Cast.toNumber(item[0]),
            Cast.toNumber(item[1]),
            Cast.toNumber(item[2])
          )
        );
      }
    }
  } else {
    console.warn(
      `Attempted to process nonexistent vertex list "${list}"`
    );
  }
  return points;
}
```

Brackets-Coder commented:
> Hey, I could not find an issue place, so I'll place it here: <img alt="image" width="712" height="388" src="https://private-user-images.githubusercontent.com/178073621/572741287-4925bc00-cf41-4a8f-afc6-9a65de916bb5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzUwODczNTQsIm5iZiI6MTc3NTA4NzA1NCwicGF0aCI6Ii8xNzgwNzM2MjEvNTcyNzQxMjg3LTQ5MjViYzAwLWNmNDEtNGE4Zi1hZmM2LTlhNjVkZTkxNmJiNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwNDAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDQwMVQyMzQ0MTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zYWFlNDZjZmQ1MzU3OGUwZjg2NGFkMjEyMjI3ODdhZTllZDY3ZWY0NTg5OGNhYTZhZGYwNTQwMWVkZTMwMjE4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._3LWpHi5omB0bzrL9xPwETDf0aC0eSXMDyxOSxpmOJw"> I guess here:
> 
> ```js
> function processVertices(list) {
>   const points = [];
>   const array = list;
>   if (array) {
>     for (let i = 0; i < array.length; i++) {
>       if (array[i] != "") {
>         const item = array[i].split(" ");
>         if (item.length !== 3) {
>           return;
>         }
>         points.push(
>           new Ammo.btVector3(
>             Cast.toNumber(item[0]),
>             Cast.toNumber(item[1]),
>             Cast.toNumber(item[2])
>           )
>         );
>       }
>     }
>   } else {
>     console.warn(
>       `Attempted to process nonexistent vertex list "${list}"`
>     );
>   }
>   return points;
> }
> ```

Thx for checking, I've identified right after looking at the image. You can report issues on the [public repository](https://github.com/Brackets-Coder/AmmoPhysics). Also, read [the documentation](https://github.com/Brackets-Coder/extensions/blob/64b95ae7a4d61fbbd9bfc384ffd48b5362e617d5/docs/MasterMath/AmmoPhysics.md) for a better explanation. The block expects a list where each item is a vertex consisting of three space-delimited coordinates. E.g.,
```json
[
  {1, 2, 3},
  {2, 3, 4},
  {4, 5, 6}
]
```

You've only put one value in each item, so splitting by spaces fails. I'll improve list validation sometime this week but for now just using the upload from OBJ block works great in my testing. Export a triangulated mesh from Blender and import the .OBJ into a list. Then just use the `import mesh from OBJ` block.

Brackets-Coder commented:
See Brackets-Coder/AmmoPhysics#10

Brackets-Coder committed:
Improve vertex list validation

Brackets-Coder commented:
@Civ3ro should be fixed in latest commit, thanks for catching the issue!

Meehdrescher commented:
Is there also going to be a separate, self-contained version of this extension? I can't get the it to load because the dependencies couldn't be fetched, the older versions that contained the dependencies worked fine

Brackets-Coder commented:
> Is there also going to be a separate, self-contained version of this extension? I can't get the it to load because the dependencies couldn't be fetched, the older versions that contained the dependencies worked fine

Probably not, a version that includes the dependencies gets very large. `Scratch.external` usually works fine in the latest version of Turbowarp, so if you're running TW Desktop try updating to the latest version or use the web editor. Also, check the JavaScript console to make sure raw.githubusercontent.com isn't blocked or something

Brackets-Coder committed:
comment grammar

Meehdrescher commented:
> 
> Probably not, a version that includes the dependencies gets very large. `Scratch.external` usually works fine in the latest version of Turbowarp, so if you're running TW Desktop try updating to the latest version or use the web editor. Also, check the JavaScript console to make sure raw.githubusercontent.com isn't blocked or something

I know it would become very large, but the size is not an issue for me and the point is to have it working offline at all times no matter what

Brackets-Coder commented:
> > Probably not, a version that includes the dependencies gets very large. `Scratch.external` usually works fine in the latest version of Turbowarp, so if you're running TW Desktop try updating to the latest version or use the web editor. Also, check the JavaScript console to make sure raw.githubusercontent.com isn't blocked or something
> 
> I know it would become very large, but the size is not an issue for me and the point is to have it working offline at all times no matter what

When the extension is merged, the URLs will be bundled into the fileso it will work offline, so the size will likely increase then. As of now it keeps the size down and keeps the code clean but I'll see what I can do to help you  (probably later when I have time)

GarboMuffin commented:
you're not supposed to touch that file manually. `npm run start` and it'll update things to match

Brackets-Coder committed:
like this?

GarboMuffin commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,3015 @@
+        }
+
+        createCompoundShape(name) {
+          name = Cast.toString(name);
```
name will always be `[object Object]`

Brackets-Coder committed:
Fix revealed issues + miscellaneous changes

Brackets-Coder commented on extensions/MasterMath/AmmoPhysics.js:
```diff
@@ -0,0 +1,3047 @@
+            Ammo.destroy(body);
+            delete bodies[name];
+          }
+        }
```
second conditional is always truthy

```suggestion
       const body = bodies[name];
       if (body) {
         world.removeRigidBody(body);
         world.removeCollisionObject(body);
         Ammo.destroy(body.getMotionState());
         Ammo.destroy(body.getCollisionShape());
         Ammo.destroy(body);
         delete bodies[name];
       }
```

Brackets-Coder committed:
Remove truthy unecessary conditional

Brackets-Coder commented:
Review maybe somebody someday eventually?
